download/tasks.py

```diff
@@ -1,44 +1,151 @@
-import glob
-from os import path
+import fnmatch
+import hashlib
+import json
+import re
+import sys
 import tempfile
+import requests
+from datetime import datetime, timedelta
+from os import path, walk
 
 from invoke import task
 
 
+STATUS_FILE_NAME = "_files_status.json"
+
+
+def raw_files_status(ctx):
+    files_status_path = path.join(ctx.cache_dir, STATUS_FILE_NAME)
+
+    if path.isfile(files_status_path):
+        with open(files_status_path) as data:
+            return json.load(data)
+    else:
+        return dict()
+
+
+def get_file_status(ctx, filename):
+    res = raw_files_status(ctx).get(filename)
+
+    if res is None:
+        res = {"last_update": None, "md5": None}
+
+    if res["last_update"] is not None:
+        res["last_update"] = datetime.utcfromtimestamp(res["last_update"])
+
+    return res
+
+
+def save_file_status(ctx, filename, status):
+    files_status_path = path.join(ctx.cache_dir, STATUS_FILE_NAME)
+
+    if status is not None:
+        status["last_update"] = status["last_update"].timestamp()
+
+    files_status = raw_files_status(ctx)
+    files_status[filename] = status
+
+    with open(files_status_path, "w") as data:
+        json.dump(files_status, data)
+
+
+def needs_to_download(ctx, filename, max_age=None, md5_url=None):
+    """
+    Check if "filename" should be downloaded because it doesn't exist or is
+    older than "max_age" or its checksum has changed.
+    """
+    if not path.isfile(filename):
+        return True
+
+    file_status = get_file_status(ctx, filename)
+
+    if max_age is not None and (
+        file_status["last_update"] is None
+        or datetime.utcnow() - file_status["last_update"] > max_age
+    ):
+        return True
+
+    if md5_url is not None:
+        res = requests.get(md5_url)
+        expt_md5 = res.text.split()[0]
+        curr_md5 = file_status["md5"]
+
+        if expt_md5 != curr_md5:
+            return True
+
+    if ctx.force_downloads:
+        print(
+            f"'force_downloads' is set to true: existing filename {filename} will be ignored",
+            file=sys.stderr,
+        )
+        return True
+
+    print(
+        f"filename '{filename}' already exists, we don't need to download it again",
+        file=sys.stderr,
+    )
+    return False
+
+
+def download_file(ctx, filename, url, max_age=None, md5_url=None):
+    if not needs_to_download(ctx, filename, max_age, md5_url):
+        return
+
+    #  Forget current informations about file in case the download fails.
+    save_file_status(ctx, filename, None)
+
+    ctx.run(f"mkdir -p {path.dirname(filename)}/")
+    ctx.run(f"wget --progress=dot:giga -O {filename} {url}")
+
+    with open(filename, "rb") as f:
+        md5 = hashlib.md5()
+        for chunk in iter(lambda: f.read(4096), b""):
+            md5.update(chunk)
+        md5 = md5.hexdigest()
```

Comment:
Maybe we should check the integrity of the downloaded file, thanks to this md5 ?

Reply:
It will now stop the execution when this happens, this is a bit brutal but at least we are sure not to miss the issue ?

Maybe we could try to download a few more times before crashing ?

Reply:
Hum yes, that would be nice. But this case should be very rare.  
Let's keep that in mind for another issue, I don't want to delay this PR for too long.